File: src/org/helllabs/android/xmp/modarchive/result/ArtistResult.java
package org.helllabs.android.xmp.modarchive.result;

import java.util.List;

import org.helllabs.android.xmp.R;
import org.helllabs.android.xmp.modarchive.Search;
import org.helllabs.android.xmp.modarchive.adapter.ArtistArrayAdapter;
import org.helllabs.android.xmp.modarchive.model.Artist;
import org.helllabs.android.xmp.modarchive.request.ArtistRequest;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;

public class ArtistResult extends Result implements ArtistRequest.OnResponseListener<List<Artist>>, ListView.OnItemClickListener {
	
	private ListView list;

	@Override
	public void onCreate(final Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.result_list);
		setupCrossfade();
		
		setTitle("Artist search");
		
		list = (ListView)findViewById(R.id.result_list);
		
		final String searchText = getIntent().getStringExtra(Search.SEARCH_TEXT);
		
		final String key = getString(R.string.modarchive_apikey);
		final ArtistRequest request = new ArtistRequest(key, "search_artist&query=" + searchText);
		request.setOnResponseListener(this).send();
		
		list.setOnItemClickListener(this);
	}
	
	@Override
	public void onResponse(final List<Artist> response) {
		final ArtistArrayAdapter adapter = new ArtistArrayAdapter(this, android.R.layout.simple_list_item_1, response);
		list.setAdapter(adapter);
		crossfade();
	}
	
	@Override
	public void onError(final Throwable error) {
		handleError(error);
	}

	@Override
	public void onItemClick(final AdapterView<?> parent, final View view, final int position, final long id) {
		final ArtistArrayAdapter adapter = (ArtistArrayAdapter)parent.getAdapter();
		final Intent intent = new Intent(this, ArtistModulesResult.class);
		intent.putExtra(Search.ARTIST_ID, adapter.getItem(position).getId());
		startActivity(intent);
	}

}

File: app/src/main/java/org/helllabs/android/xmp/modarchive/result/ArtistResult.kt
package org.helllabs.android.xmp.modarchive.result

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.AdapterView
import android.widget.ListView
import android.widget.TextView
import org.helllabs.android.xmp.BuildConfig
import org.helllabs.android.xmp.R
import org.helllabs.android.xmp.modarchive.Search
import org.helllabs.android.xmp.modarchive.adapter.ArtistArrayAdapter
import org.helllabs.android.xmp.modarchive.request.ArtistRequest
import org.helllabs.android.xmp.modarchive.request.ModArchiveRequest
import org.helllabs.android.xmp.modarchive.request.ModArchiveRequest.OnResponseListener
import org.helllabs.android.xmp.modarchive.response.ArtistResponse
import org.helllabs.android.xmp.modarchive.response.HardErrorResponse
import org.helllabs.android.xmp.modarchive.response.ModArchiveResponse
import org.helllabs.android.xmp.modarchive.response.SoftErrorResponse
import java.io.UnsupportedEncodingException

class ArtistResult : Result(), OnResponseListener, AdapterView.OnItemClickListener {
    private var list: ListView? = null
    private var errorMessage: TextView? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.result_list)
        setupCrossfade()
        setTitle(R.string.search_artist_title)
        list = findViewById<View>(R.id.result_list) as ListView
        list!!.onItemClickListener = this
        errorMessage = findViewById<View>(R.id.error_message) as TextView
        val searchText = intent.getStringExtra(Search.Companion.SEARCH_TEXT)
        val key = BuildConfig.API_KEY
        try {
            val request = ArtistRequest(key, ModArchiveRequest.Companion.ARTIST, searchText)
            request.setOnResponseListener(this).send()
        } catch (e: UnsupportedEncodingException) {
            handleQueryError()
        }
    }

    override fun onResponse(response: ModArchiveResponse) {
        val artistList = response as ArtistResponse
        val adapter = ArtistArrayAdapter(this, android.R.layout.simple_list_item_1, artistList.list)
        list!!.adapter = adapter
        if (artistList.isEmpty) {
            errorMessage!!.setText(R.string.search_no_result)
            list!!.visibility = View.GONE
        }
        crossfade()
    }

    override fun onSoftError(response: SoftErrorResponse) {
        val errorMessage = findViewById<View>(R.id.error_message) as TextView
        errorMessage.text = response.message
        list!!.visibility = View.GONE
        crossfade()
    }

    override fun onHardError(response: HardErrorResponse) {
        handleError(response.error)
    }

    override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        val adapter = parent.adapter as ArtistArrayAdapter
        val intent = Intent(this, ArtistModulesResult::class.java)
        intent.putExtra(Search.Companion.ARTIST_ID, adapter.getItem(position)!!.id)
        startActivity(intent)
        overridePendingTransition(R.anim.slide_in_right, R.anim.slide_out_left)
    }
}
